arch/arm/src/nrf52/nrf52_sdc.c

```diff
@@ -0,0 +1,609 @@
+ ****************************************************************************/
+
+static const struct bt_driver_s g_bt_driver =
+{
+  .head_reserve = 0,
+  .open         = bt_open,
+  .send         = bt_hci_send
+};
+
+static const mpsl_clock_lfclk_cfg_t g_clock_config =
+{
+  .source                   = MPSL_CLOCK_LF_SRC_XTAL,
+  .rc_ctiv                  = 0,
+  .rc_temp_ctiv             = 0,
+  .accuracy_ppm             = CONFIG_NRF52_SDC_CLOCK_ACCURACY,
+  .skip_wait_lfclk_started  = false
+};
+
+static struct nrf52_sdc_dev_s g_sdc_dev;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: bt_open
+ ****************************************************************************/
+
+static int bt_open(FAR const struct bt_driver_s *btdev)
+{
+  return 0;
+}
+
+/****************************************************************************
+ * Name: bt_open
+ ****************************************************************************/
+
+static int bt_hci_send(FAR const struct bt_driver_s *btdev,
+                       FAR struct bt_buf_s *buf)
+{
+  int ret = OK;
+  irqstate_t flags;
+
+  /* Pass HCI CMD/DATA to SDC */
+
+  if (buf->type == BT_CMD)
+    {
+      struct bt_hci_cmd_hdr_s *cmd = (struct bt_hci_cmd_hdr_s *)buf->data;
+
+      wlinfo("passing CMD %d to softdevice\n", cmd->opcode);
+
+      /* Ensure non-concurrent access to SDC operations */
+
+      flags = enter_critical_section();
+
+      nxsem_wait_uninterruptible(&g_sdc_dev.exclsem);
+
+      if (sdc_hci_cmd_put(buf->data) < 0)
+        {
+          wlerr("sdc_hci_cmd_put() failed\n");
+          ret = -EIO;
+        }
+
+      nxsem_post(&g_sdc_dev.exclsem);
+
+      leave_critical_section(flags);
+
+      work_queue(LPWORK, &g_sdc_dev.work, on_hci_worker, NULL, 0);
+    }
+  else if (buf->type == BT_ACL_OUT)
+    {
+      wlinfo("passing ACL to softdevice\n");
+
+      /* Ensure non-concurrent access to SDC operations */
+
+      flags = enter_critical_section();
+
+      nxsem_wait_uninterruptible(&g_sdc_dev.exclsem);
+
+      if (sdc_hci_data_put(buf->data) < 0)
+        {
+          wlerr("sdc_hci_data_put() failed\n");
+          ret = -EIO;
+        }
+
+      nxsem_post(&g_sdc_dev.exclsem);
+
+      leave_critical_section(flags);
+
+      work_queue(LPWORK, &g_sdc_dev.work, on_hci_worker, NULL, 0);
+    }
+
+  if (ret < 0)
+    {
+      wlerr("bt_hci_send() failed: %d\n", ret);
+      return ret;
+    }
+  else
+    {
+      return buf->len;
+    }
+}
+
+/****************************************************************************
+ * Name: sdc_fault_handler
+ ****************************************************************************/
+
+static void sdc_fault_handler(const char *file, const uint32_t line)
+{
+  _alert("SoftDevice Controller Fault\n");
```

Comment:
I think we should make this _alert wlerr we reserve _alert for very few places, and you will already get the context from the assert.

Reply:
The problem is that the call only passes on the file/line received from SDC/MPSL which does not give context (the file field appears to be a number sometimes, seems filename is ofuscated). And I wanted to avoid too much head-scratching if this happens but wireless errors are disabled.